### ISSUE TITLE
How to profile async code?

### ISSUE THREAD

**Issue by careless6666** [grafana/pyroscope-dotnet#50]
If we check documentation
https://grafana.com/docs/pyroscope/latest/configure-client/language-sdks/dotnet/

now we can profile only sync code

```
var labels = Pyroscope.LabelSet.Empty.BuildUpon()
    .Add("key1", "value1")
    .Build();
Pyroscope.LabelsWrapper.Do(labels, () =>
{
  SlowCode();
});
```

Do you have plans for add DoAsync version?

**Comment by craigbehnke** [comment]
I was just about to sit down and integrate Pyroscope into my server code when I ran into this issue myself.

My (potentially misguided) inclination would be to modify [Pyroscope/Pyroscope/LabelsWrapper.cs](https://github.com/grafana/pyroscope-dotnet/blob/0beaaf93fdfe8a9272fc05c4b793d455e74837ad/Pyroscope/Pyroscope/LabelsWrapper.cs) so that it has the following `DoAsync` method (I'm including the entire file so that the new and old functions can be easily compared):

```c#
namespace Pyroscope;

public static class LabelsWrapper
{
    public static void Do(LabelSet labels, Action a)
    {
        try
        {
            labels.Activate();
            a.Invoke();
        }
        finally
        {
            if (labels.Previous == null)
            {
                Profiler.Instance.ClearDynamicTags();
            }
            else
            {
                labels.Previous.Activate();
            }
        }
    }

    /// <summary>
    /// New method to handle async actions
    /// </summary>
    public static async Task DoAsync(LabelSet labels, Func<Task> a)
    {
        try
        {
            labels.Activate();
            await a.Invoke();
        }
        finally
        {
            if (labels.Previous == null)
            {
                Profiler.Instance.ClearDynamicTags();
            }
            else
            {
                labels.Previous.Activate();
            }
        }
    }
}
```

@korniltsev, is there a technical reason this has not been implemented (i.e. the profiler cannot handle async methods or something like that), or might the method above work?

Alternatively, is there a work-around that would allow us to profile async code without modifying the existing codebase?

**Comment by korniltsev** [comment]
No, current labels implementation will not work with async code, they are per OS thread. 
We'd  need something more specific to .net async functionality to make it work

**Comment by craigbehnke** [comment]
Is there currently a non-labels way to get request-level profiling data?

And thank you for the fast response!

**Comment by korniltsev** [comment]
> Is there currently a non-labels way to get request-level profiling data?
Unfortunately no

**Comment by RedEye-Developers** [comment]
any update for `DoAsync` ?

**Comment by Dawid33** [comment]
I implemented async profiling functionality with claude in the native profiler. Its a bit of a slop pr right now but its functional. Is there any interest in me continuing this work, cleaning it up etc? Is it still the case that there is no plan to support async? #433

**Comment by korniltsev-grafanista** [comment]
I have not looked into the code yet.
That is quite a big change.
It would be really nice if it landed upstream https://github.com/DataDog/dd-trace-dotnet first. Have you considered contributing it there? 
Is there any way we can split the change in smaller easier to review chunks?

**Comment by Dawid33** [comment]
Hi @korniltsev-grafanista, I haven't considered upstream because we don't use any of datadog's stuff at work (I can't test with our stuff) but yeah I'll look into what upstream thinks of this sort of feature first in that case. That pr is several  big changes rolled into one (async labels, async scopes, fixing the stack to stitch together async code to the right parent, removing async statemachine code to clean up the stack etc) so It can definitely be split up. I'm glad there is interest anyhow. I'll get back to you here once I have a chance to check with them.
